### PR TITLE
Clean up BIR after codegen phase

### LINKED ISSUE #41384: [Task]: Optimise ballerina bir storing logic
### Description

$Subject requires for reduce memory during ballerina compilation 

### Describe your task(s)

_No response_

### Related area

-> Runtime

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by Nadeeshan96**
$title
This is to reduce memory taken during compilation.
Fixes #41384

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmDesugarPhase.java [hunk 1]
```diff
@@ -160,275 +154,120 @@ private static void rewriteRecordInitFunction(BIRFunction func, BRecordType reco
...
-        return encodedIdString;
-    }
-
-    private static Name getInitialIdString(Name encodedIdString, HashMap<String, String> encodedVsInitialIds) {
-        String initialString = encodedVsInitialIds.get(encodedIdString.value);
-        if (initialString != null) {
-            return Names.fromString(initialString);
-        }
-        return encodedIdString;
+    private static String encodeNonFunctionIdentifier(String identifier) {
```
Comment (HindujaB) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmDesugarPhase.java:270]: As this function only contains a static method call, we can remove this and use the `Utils.encodeNonFunctionIdentifier(identifier)` directly.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmDesugarPhase.java [hunk 2]
```diff
@@ -160,275 +154,120 @@ private static void rewriteRecordInitFunction(BIRFunction func, BRecordType reco
...
-                }
-            }
-            if (bType.tag == TypeTags.RECORD) {
-                BRecordType recordType = (BRecordType) bType;
-                for (BField field : recordType.fields.values()) {
-                    field.name = getInitialIdString(field.name, encodedVsInitialIds);
-                }
-            }
-        }
+    private static String encodeFunctionIdentifier(String identifier) {
```
Comment (HindujaB) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmDesugarPhase.java:266]: As this function only contains a static method call, we can remove this and use the `Utils.encodeFunctionIdentifier(identifier)` directly.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmDesugarPhase.java [hunk 3]
```diff
@@ -160,275 +154,120 @@ private static void rewriteRecordInitFunction(BIRFunction func, BRecordType reco
...
-    private static void encodeTypeDefIdentifiers(List<BIRTypeDefinition> typeDefs,
-                                                 HashMap<String, String> encodedVsInitialIds) {
+    private static void encodeTypeDefIdentifiers(List<BIRTypeDefinition> typeDefs) {
         for (BIRTypeDefinition typeDefinition : typeDefs) {
-            typeDefinition.type.tsymbol.name = Names.fromString(encodeNonFunctionIdentifier(
-                    typeDefinition.type.tsymbol.name.value, encodedVsInitialIds));
-            typeDefinition.internalName =
-                    Names.fromString(encodeNonFunctionIdentifier(typeDefinition.internalName.value,
-                            encodedVsInitialIds));
+            Name encodedName = Names.fromString(encodeNonFunctionIdentifier(
```
Comment (HindujaB) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmDesugarPhase.java:172]: We can also move the `Names.fromString` part inside the `encodeNonFunctionIdentifier` and `encodeFunctionIdentifier` private methods

**Review by HindujaB — APPROVED** [review]
LGTM

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmDesugarPhase.java [hunk 4]
```diff
@@ -160,275 +154,119 @@ private static void rewriteRecordInitFunction(BIRFunction func, BRecordType reco
...
-            function.funcName = getInitialIdString(function.funcName, encodedVsInitialIds);
+            function.funcName = encodeFunctionIdentifier(function.funcName.value);
         }
     }
 
-    private static void replaceEncodedGlobalVariableIdentifiers(List<BIRNode.BIRGlobalVariableDcl> globalVars,
-                                                                HashMap<String, String> encodedVsInitialIds) {
+    private static void encodeGlobalVariableIdentifiers(List<BIRNode.BIRGlobalVariableDcl> globalVars) {
         for (BIRNode.BIRGlobalVariableDcl globalVar : globalVars) {
+            // globalVar should not be null, check and remove this
```
Comment (gabilang) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmDesugarPhase.java:251]: Are we going to address this in the same PR? If not it's better to create a separate task and mention it as TODO. I'll be easier to track.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmDesugarPhase.java [hunk 5]
```diff
@@ -160,275 +154,119 @@ private static void rewriteRecordInitFunction(BIRFunction func, BRecordType reco
         func.localVars = updatedLocalVars;
     }
 
-    private static BIRFunctionParameter getFunctionParam(BIRFunctionParameter localVar) {
-        if (localVar == null) {
-            throw new BLangCompilerException("Invalid function parameter");
-        }
-
-        return localVar;
+    static void encodeModuleIdentifiers(BIRNode.BIRPackage module) {
```
Comment (gabilang) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmDesugarPhase.java:157]: Is an access modifier not necessary here?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmPackageGen.java [hunk 6]
```diff
@@ -842,7 +842,9 @@ private boolean listenerDeclarationFound(BPackageSymbol packageSymbol) {
                 if (importPkgSymbol == null) {
                     continue;
                 }
-                return listenerDeclarationFound(importPkgSymbol);
+                if (listenerDeclarationFound(importPkgSymbol)) {
+                    return true;
+                }
```
[the remark this answers is not in the document]
Comment (Nadeeshan96) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmPackageGen.java:847]: The earlier logic is not correct. It just check the first non null `importPkgSymbol` return whether if it has a listener as the whole answer. But we should return here whether a listener is there in all the imported packages.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/CodeGenerator.java [hunk 7]
```diff
@@ -95,16 +91,47 @@ private CompiledJarFile generate(BPackageSymbol packageSymbol) {
...
 
         //Rewrite identifier names with encoding special characters
-        HashMap<String, String> originalIdentifierMap = JvmDesugarPhase.encodeModuleIdentifiers(packageSymbol.bir);
+        JvmDesugarPhase.encodeModuleIdentifiers(packageSymbol.bir);
 
         // TODO Get-rid of the following assignment
         packageSymbol.compiledJarFile = jvmPackageGen.generate(packageSymbol.bir, true);
-
-        //Revert encoding identifier names
-        JvmDesugarPhase.replaceEncodedModuleIdentifiers(packageSymbol.bir, originalIdentifierMap);
```
Comment (Nadeeshan96) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/CodeGenerator.java:104]: We can't remove this line 104. I will send a commit to revert it. Though all the tests when I compile some packages for the second time, an error is thrown at BIRSymbolEnter. We need to either change BIRSymbolEnter + BIRBinaryWriter or replace the encoded identifiers.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmDesugarPhase.java (hunk 5 above)
Comment (Nadeeshan96) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmDesugarPhase.java:157]: It was previously there. Hence, that is okay kept it.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmDesugarPhase.java (hunk 4 above)
Comment (Nadeeshan96) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmDesugarPhase.java:251]: Removed this.

**Comment by Nadeeshan96** [comment]
Sent a new PR for this: https://github.com/ballerina-platform/ballerina-lang/pull/41510. Hence closing this. PR review comments are resolved in the new PR.

**Pull request closed without merging**